Add support for --null-input option

bfrg opened the pull request:
Fixes #2.

There is one small limitation: When `-n` or `--null-input` are passed to `:Jqplay` or `:JqplayScratch`, the filter and output buffers are always opened on a new tab page.

bfrg committed:
Don't use any input buffer when -n option passed

When the -n or --null-input options are passed to jq, no input buffer is
read. The following commit adds support for this.

When :Jqplay is run with -n/--null-input, the jq filter and json output
buffer are always opened in a new tab page. The current json buffer is
ignored.

This partially solves #2.

TODO: We still need a global command that can be run from any buffer,
not just json.

bfrg committed:
Rename jqplay#scratch() function

bfrg committed:
Add global command :JqplayScratch

:JqplayScratch can be called from any filetype. It always starts an
interactive session in a new tab page using a new json scratch buffer,
unless the -n/--null-input options are passed. In this case, the command
is identical to running ":Jqplay -n".

odnoletkov commented on autoload/jqplay.vim:
```diff
@@ -59,36 +59,48 @@ function! jqplay#scratch(mods, args) abort
     endif
 
+    " If -n/--null-input is passed we won't need an input buffer for jq
+    let null_input = a:args =~# '-n\|--null-input' ? 1 : 0
```
Short-form options can be combined in a single argument: `jq -Sn <filter>`. Something like `'-\S*n\|--null-input'` could take care of that

bfrg replied:
I will add support for `--raw-input`. It's probably best to make `:Jqplay` a global command, and then just do a check if `-n` or/and `-R` have been passed. A command like `:JqplayScratch` will nevertheless be required in case the current buffer is json but someone wants to start a new session with a new json scratch buffer and not the current one.

Can you think of another case where `-n` is required? The issue you linked above suggests that it should be added whenever the `inputs` function is used.

bfrg replied:
This means, when I run
```bash
cat foo.json | jq -n 'input'
```
the file `foo.json` is read by `jq`, but when I omit the `input` function, it won't be parsed:
```bash
cat foo.json | jq -n '.'
```
Ouput: `null`

odnoletkov reviewed (commented):
One more thought about `--null-input` came to mind when testing this. I remember another use case for it in jq. When using `inputs` it prevents the first input to be automatically consumed: https://github.com/stedolan/jq/issues/1956. It is useful for example for feeding line-based input into jq:
```
$ cat file | jq -nR '[inputs] | …'
```

So in practice `-n` doesn't *always* mean there is no input file.

Currently it is not a big deal because `:Jqplay` could only be invoked from a json buffer – which naturally implies a single input case. But feeding non-json buffers to jqplay seems like a possible idea for the future? So maybe it is better to handle `-n`/`--null-input` only in `:JqplayScratch` – and pass the *optional* input buffer to jqplay#start explicitly?

bfrg committed:
Support combined short-form options

bfrg committed:
Add support for --raw-input, make :Jqplay global

Make :Jqplay globally accessible so that it can be run from any filetype
not just json. This is useful when the -R/--raw-input option is passed.

The plugin will check if --null-input and --raw-input are used together.
When both options are passed to :Jqplay, the input buffer nevertheless
needs to be passed to jq despite the -n option. For an example, see
https://github.com/stedolan/jq/issues/1956

odnoletkov commented:
Just tested the updates – great work! Global `:Jqplay` is very useful and seems like a natural fit.

The only minor improvement I see is `-n` is not always strictly tied to `-R`. It can be used in practice w/o it. For example consider file with numbers only:
```
$ seq 10 | vim -
```
You don't have to use `-R`/`--raw-input` to feed this buffer to jq because numbers don't need to be converted to strings to be used in jq. So I would expect `:Jqplay -n` to work – i.e. create a normal session where i can start with the `[inputs]` filter. But currently jqplay will create input-less session because there is no `-R`. In other words there are valid use cases for `-n` *with* the input buffer and *without* `-R`.

Like I suggested above – maybe `-n` should be treated specially – and create an input-less session – only in `:JqplayScratch`? It looks like it would also simplify the code.

bfrg commented:
It is indeed a limitation. And I don't like the current implementation, I would need to refactor it anyway.

> Like I suggested above – maybe -n should be treated specially – and create an input-less session – only in :JqplayScratch? It looks like it would also simplify the code.

Yes, this would certainly make the implementation simpler. 
I like the idea of always using the current buffer as input when `:Jqplay -n` is called. There's one minor issue though: We don't know in advance whether or not the user will use the `inputs` filter. Imagine someone doesn't use the `inputs` filter and so doesn't care about the current buffer. This is fine since `jq` will never parse it. But if the current buffer is used as input when `:Jqplay -n` is called, then this buffer will have to stick around as long as the session is open. If you delete or wipe out that buffer, the session will be closed automatically.

>  [...] create an input-less session – only in :JqplayScratch? [...]

If `:JqplayScratch -n` always creates an input-less session, then what if someone wants to open a new session with `:JqplayScratch -n` and create the sequence of numbers in the newly created scratch buffer (instead of `seq 10 | vim -`)? This wouldn't work then. Instead you would have to open a new scratch buffer manually: `:tabnew my_seq`, _insert number_ etc, run `:Jqplay -n`. 

We could also do it the other way around: Running `:Jqplay! -n` with a bang means _I will use the `inputs` filter_, hence pass the current buffer to `jq`. While `:Jqplay -n` always creates an input-less session.

Personally I have never used the `inputs` filter nor the `-n` option. So I can't tell which case is more common `-n` without the `inputs` filter, or `-n` with the `inputs` filter.

bfrg commented:
To me it seems like it's less common to pass the `--null-input` option to `jq` and still use an input file.
The man page even says:
```
--null-input/-n:
    Don´t read any input at all! Instead, the filter is run once using null as the input.
```
 So `:Jqplay! -n` would make more sense. The bang means _read the current buffer!_ Then when you know that you will use the `inputs` filter, you will have to run `:Jqplay! -n`.

odnoletkov commented:
Agree on all the above. It looks like the root of the problem is our assumption that we can derive users intent – using input buffer or not – just from flags. While in practice flags story is more complex.

What about different, simpler approach:
- `:Jqplay` always uses the current buffer as a source
- `:JqlayScratch` creates an empty json buffer and uses it as a source
- `:JqplayScratch!` creates a source-less session (no source buffer and `-n` is forced)

Then we don’t need to know/deal with flags. Users who use `-n` will be able to do so – and behavior will be the same as on the command line.

bfrg commented:
I'm closing this one see #4 instead.
